File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
  if (n === 1) return ['()'];
  let res = [];
  helper(n, n, ''); 
  return res;
  
  //Helper function
  function helper(open, close, combo) {
    if (close < open) return;
    if (close === 0 && open === 0) return res.push(combo);
    if (open > 0) helper(open-1, close, combo+'(');
    if (close > 0) helper(open, close-1, combo+')');
  }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
      if n == 1: 
        return ['()']
      res = []
      self.helper(n, n, '', res)
      return res
  
    def helper(self, open, close, str, result):
      if close < open:
        return
      if open == 0 and close == 0:
        result.append(str)
        return
    
      if open > 0:
        self.helper(open - 1, close, f'{str}(', result)
        
      if close > 0:
        self.helper(open, close - 1, f'{str})', result)